Update kustomization to use params env and yml on notebook-controller and odh-notebook-controller

atheo89 opened the pull request:
Related to: https://issues.redhat.com/browse/RHOAIENG-9008

## Description
Sync manifests among upstream `v1.7-branch` and downstream `master` branches

Note: For reference, the parametrization of the manifests incorporated downstream via this PR is as follows::  https://github.com/red-hat-data-services/kubeflow/pull/29 

## How has been tested

**Evaluate notebook-controller deployment by running the following:**
`$ cd components/notebook-controller`
`$ kustomize build config/overlays/openshift`

Check on the image on the `notebook-controller-deployment` deployment that matched with the 
`image: quay.io/opendatahub/kubeflow-notebook-controller:1.7-35b81f5`

**Evaluate odh-notebook-controller deployment by running the following:**
`$ cd components/odh-notebook-controller`
`$ kustomize build config/base`

Check on the image on the deployment of `odh-notebook-controller-manager` that matched with the 
`image: quay.io/opendatahub/kubeflow-notebook-controller:1.7-35b81f5`


## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

jiridanek commented:
I'm looking at a diff between your branch and red-hat-data-services/kubeflow:main, and I see lots of noise in the diff. And that's despite your PR branch is in sync with opendatahub-io/notebooks:v1.7-branch and individual files that show differences in the diff view are actually much more similar in reality

`components/odh-notebook-controller/config/base/kustomization.yaml` looks very different in the diff

https://github.com/atheo89/kubeflow/compare/work-on-kustomize...red-hat-data-services:kubeflow:master#diff-b02147f941d8f6f01343f0167180165c33c9469e108f7f66d063bb4a4aa6b540

but looking at the actual file in both repos, it only differs in a few blank lines

https://github.com/atheo89/kubeflow/blob/work-on-kustomize/components/odh-notebook-controller/config/base/kustomization.yaml

https://github.com/red-hat-data-services/kubeflow/blob/master/components/odh-notebook-controller/config/base/kustomization.yaml

I'm still trying to understand if I am using GitHub wrong. I may have to checkout both locally and run diff on my machine.

jiridanek commented:
Ok, local diff works much better; here are the relevant differences. I wrote down how I got the diff on https://stackoverflow.com/questions/1968512/getting-the-difference-between-two-repositories/78734836#78734836

## First differences,

![image](https://github.com/opendatahub-io/kubeflow/assets/442720/bd4c1a67-85a8-4a59-a81a-66a05276f0c8)

downstream has the operator image parameterized, and references the ose-oauth-proxy by hash

https://github.com/atheo89/kubeflow/blob/work-on-kustomize/components/odh-notebook-controller/config/manager/manager.yaml#L24C1-L24C68

## Second differences

the manager-openshift-patch.yaml file is different

![image](https://github.com/opendatahub-io/kubeflow/assets/442720/a89b230e-ab14-4a9e-a096-0f834e77221f)

atheo89 commented:
Thank for the review! 

> My understanding of this task is that manifests between opendatahub-io and red-hat-data-services repos should be made identical, with the only difference of image hashes in the .env files. If you think the scope is smaller than that, I can accept it, just that we are on the same page about this.

You mean that we have to apply this sync in all manifest files to be identical with downstream?  if so, i guess i can do that. I had the impression according to the issue description that we have to do it on the  https://github.com/opendatahub-io/kubeflow/blob/v1.7-branch/components/notebook-controller/config/overlays/openshift/kustomization.yaml however i expand it also to odh-notebook-controller

jiridanek commented:
Description on the Jira first says

> The current opendatahub(upstream) kubeflow and red-hat-data-services(downstream) kubeflow are out-of-sync based on the way manifests configs are maintained. This Jira is to have them in sync so that automation and CI checks don't need to be maintained separately. 

but then the Dev section below is more selective, as you say.

atheo89 commented:
Give me some time to discuss and groom this further with Harshad and i will check it again

atheo89 commented:
\hold

harshad16 reviewed (changes requested):
This changes are good, however it is missing the referencing of the image.
as the `image` substitution is changed from kustomize image, to config-param method, we need to invoke the variable in deployment, for the substitution to happen.

for notebook-controller:
- Following change is needed: https://github.com/red-hat-data-services/kubeflow/blob/750bf133f7b51f7d319d35389de06eadf9bb3eed/components/notebook-controller/config/overlays/openshift/manager_openshift_patch.yaml#L35

For odh-notebook-controller:
- https://github.com/red-hat-data-services/kubeflow/blob/750bf133f7b51f7d319d35389de06eadf9bb3eed/components/odh-notebook-controller/config/manager/manager.yaml#L24
- https://github.com/red-hat-data-services/kubeflow/blob/master/components/odh-notebook-controller/config/manager/params.yaml

Please take a look.

jiridanek commented:
And then, with the changes Harshad described, you'd probably have to do also this, for ci to keep working https://github.com/red-hat-data-services/kubeflow/pull/62 (it's also suggested in the QA section of the Jira issue)

atheo89 commented:
Thank you @harshad16 and @jiridanek for your review :slightly_smiling_face: 

I have addressed your comments, and the PR is now ready for another round of review.

jiridanek commented:
> the ci doesn't check the images of the controller generated of this PR

can you elaborate? If the problem you see that in the odh- case the integration test is not using the image that was just built, then that's a regression compared to `v1.7-branch`, where the just built image gets loaded

* https://github.com/opendatahub-io/kubeflow/actions/runs/9954776876/job/27501154057#step:13:27
* https://github.com/opendatahub-io/kubeflow/actions/runs/9954776854/job/27501154411#step:13:30
** on v1.7-branch, this tests the new image https://github.com/opendatahub-io/kubeflow/actions/runs/9679858685/job/26706724631#step:13:30

atheo89 commented:
Okkk, so the `localhost/notebook-controller:integration-test` is what built based on the whatever changes of this PR right?

jiridanek commented on .github/workflows/odh_notebook_controller_integration_test.yaml:
```diff
@@ -82,12 +82,14 @@ jobs:
+          # export CURRENT_NOTEBOOK_IMG=quay.io/opendatahub/odh-notebook-controller:latest
+          # export PR_NOTEBOOK_IMG=localhost/${{env.IMG}}:${{env.TAG}}
+          # kustomize edit set image ${CURRENT_NOTEBOOK_IMG}=${PR_NOTEBOOK_IMG}
+          sed -i "s/quay.io\/opendatahub\/odh-notebook-controller:stable/localhost\/${{env.IMG}}:${{env.TAG}}/g" params.env
```
you have different tag in your params file here, they need to match; or, even better, I would not use sed, but echo >, to rewrite the file entirely with the new img

```suggestion
          echo "odh-kf-notebook-controller-image=localhost\/${{env.IMG}}:${{env.TAG}}" > params.env
```

atheo89 replied:
not sure about this, however i will push your suggestion to check it

jiridanek replied:
it's never good when we have ci checks written in such a way that one can't run them on local machine easily ;(

atheo89 replied:
https://github.com/opendatahub-io/kubeflow/actions/runs/9955370866/job/27502971667?pr=364 

```
Run set -x
+ cd components/odh-notebook-controller/config/base
+ kubectl create ns opendatahub
namespace/opendatahub created
+ echo 'odh-kf-notebook-controller-image=localhost\/odh-notebook-controller:integration-test'
+ oc apply -f -
+ cat
configmap/notebook-controller-culler-config created
+ cat
+ kubectl apply -f -
secret/odh-notebook-controller-webhook-cert created
+ kustomize build .
+ sed 's/imagePullPolicy: Always/imagePullPolicy: IfNotPresent/g'
+ kubectl apply -f -
Error: field specified in var '{odh-notebook-controller-image ~G_v1_ConfigMap {data.odh-notebook-controller-image}}' not found in corresponding resource
error: no objects passed to apply
Error: Process completed with exit code 1.
```

jiridanek replied:
guess you were right to doubt this, `\/` is escape for the sed command, without sed it will be just `/`

```suggestion
          echo "odh-kf-notebook-controller-image=localhost/${{env.IMG}}:${{env.TAG}}" > params.env
```

jiridanek commented on .github/workflows/odh_notebook_controller_integration_test.yaml:
```diff
@@ -82,12 +82,14 @@ jobs:
-          export CURRENT_NOTEBOOK_IMG=quay.io/opendatahub/odh-notebook-controller:latest
-          export PR_NOTEBOOK_IMG=localhost/${{env.IMG}}:${{env.TAG}}
-          kustomize edit set image ${CURRENT_NOTEBOOK_IMG}=${PR_NOTEBOOK_IMG}
+          # export CURRENT_NOTEBOOK_IMG=quay.io/opendatahub/odh-notebook-controller:latest
+          # export PR_NOTEBOOK_IMG=localhost/${{env.IMG}}:${{env.TAG}}
+          # kustomize edit set image ${CURRENT_NOTEBOOK_IMG}=${PR_NOTEBOOK_IMG}
```
can you delete the commented out code?

jiridanek commented on components/notebook-controller/config/overlays/openshift/kustomization.yaml:
```diff
@@ -8,17 +8,27 @@ commonLabels:
   app.kubernetes.io/part-of: odh-notebook-controller
```
downstream does not have `namespace:` on line 6, is that necessary to have it hardcoded here, or should the operator be responsible for deciding the namespace depending on user configuration?

atheo89 replied:
Hmm not sure, probably there is some reason @harshad16 wdyt? 
Upstream: https://github.com/opendatahub-io/kubeflow/blob/v1.7-branch/components/notebook-controller/config/overlays/openshift/kustomization.yaml#L6 
Downstream: https://github.com/red-hat-data-services/kubeflow/blob/master/components/notebook-controller/config/overlays/openshift/kustomization.yaml

harshad16 replied:
i m not aware, let ask this to operator team.
for now, lets keep this as it is.

jiridanek commented on components/notebook-controller/config/overlays/openshift/params.env:
```diff
@@ -0,0 +1 @@
+odh-kf-notebook-controller-image=quay.io/opendatahub/kubeflow-notebook-controller:1.7-35b81f5
```
who (and when) is responsible for updating the image tag?

atheo89 replied:
The person that do the release. This procedure take place here for example https://github.com/opendatahub-io/kubeflow/pull/357 

I have open an issue to automate this https://issues.redhat.com/browse/RHOAIENG-9848

jiridanek commented on components/odh-notebook-controller/config/manager/manager.yaml:
```diff
@@ -21,7 +21,7 @@ spec:
+          image: $(odh-notebook-controller-image)
           imagePullPolicy: Always
           command:
             - /manager
```
on the line below this one, in downstream we have 

```
          args: ["--oauth-proxy-image", "registry.redhat.io/openshift4/ose-oauth-proxy@sha256:4bef31eb993feb6f1096b51b4876c65a6fb1f4401fee97fa4f4542b6b7c9bc46"]

```

is there a reason to use tag here and hash in downstream? why not use hash everywhere, and put the corresponding tag as comment for easier reference?

atheo89 replied:
Also in addition to this maybe we have to update it as well

jiridanek replied:
sure, but probably not right now, if it's touched, who knows what may break

atheo89 replied:
Opened a tracking issue: https://issues.redhat.com/browse/RHOAIENG-9935

jiridanek commented on .github/workflows/odh_notebook_controller_integration_test.yaml:
```diff
@@ -82,12 +82,10 @@ jobs:
-          export CURRENT_NOTEBOOK_IMG=quay.io/opendatahub/odh-notebook-controller:latest
-          export PR_NOTEBOOK_IMG=localhost/${{env.IMG}}:${{env.TAG}}
-          kustomize edit set image ${CURRENT_NOTEBOOK_IMG}=${PR_NOTEBOOK_IMG}
+          echo "odh-kf-notebook-controller-image=localhost\/${{env.IMG}}:${{env.TAG}}" > params.env
```
guess you were right to doubt this, `\/` is escape for the sed command, without sed it will be just `/`

```suggestion
          echo "odh-kf-notebook-controller-image=localhost/${{env.IMG}}:${{env.TAG}}" > params.env
```

atheo89 replied:
I don't know how to test this workflow locally :rofl:  so i will push once more your suggestion to see the result

jiridanek commented on components/odh-notebook-controller/config/manager/kustomization.yaml:
```diff
@@ -4,3 +4,5 @@ kind: Kustomization
   - manager.yaml
   - service.yaml
+configurations:
+- params.yaml
```
missing blank line at the end of file

jiridanek commented on components/notebook-controller/config/overlays/openshift/params.env:
```diff
@@ -0,0 +1 @@
+odh-kf-notebook-controller-image=quay.io/opendatahub/kubeflow-notebook-controller:1.7-35b81f5
```
So when we previously referenced the image as

```
image: quay.io/opendatahub/odh-notebook-controller:latest
```

that meant that whatever version of odh user installed, they would get the latest controller image we built?

does the odh operator uses some mechanism to set operator image hashes? the rhoai operator certainly does something like that

jstourac replied:
> image: quay.io/opendatahub/odh-notebook-controller:latest

I didn't check the code thoroughly, but you mean, we referenced also an image from a different repository before?

jiridanek replied:
This is just the https://issues.redhat.com/browse/RHOAIENG-9234 problem, nothing new.

harshad16 replied:
>that meant that whatever version of odh user installed, they would get the latest controller image we built?
`Latest` was the fall back, it like if kustomize fails to patch the image, it would just stay `latest`.
in previous version, kustomize image was been used.

>does the odh operator uses some mechanism to set operator image hashes? the rhoai operator certainly does something like that

rhoai operator switch image based on the sha give from builds.
perhaps worth check with operator team.

jiridanek commented on .github/workflows/odh_notebook_controller_integration_test.yaml:
```diff
@@ -82,12 +82,10 @@ jobs:
-          export CURRENT_NOTEBOOK_IMG=quay.io/opendatahub/odh-notebook-controller:latest
-          export PR_NOTEBOOK_IMG=localhost/${{env.IMG}}:${{env.TAG}}
-          kustomize edit set image ${CURRENT_NOTEBOOK_IMG}=${PR_NOTEBOOK_IMG}
+          echo "odh-kf-notebook-controller-image=localhost/${{env.IMG}}:${{env.TAG}}" > params.env
```
```suggestion
          echo "odh-notebook-controller-image=localhost/${{env.IMG}}:${{env.TAG}}" > params.env
```

-kf- is the other controller 😮‍💨

jiridanek commented:
/LGTM from me, we have some unresolved questions and followup issues in Jira, but that's just that, followup

harshad16 reviewed (approved):
/lgtm
/approve

thanks :+1:

jiridanek commented:
/cherrypick stable

jiridanek commented:
/override "Code static analysis / govulncheck (components/notebook-controller)"

jiridanek commented:
/override "govulncheck (components/notebook-controller)"

jiridanek commented:
/override "govulncheck (components/odh-notebook-controller)"

Vulnerabille go deps are already present on base branch
